binderhub/build.py

```diff
@@ -49,133 +52,62 @@ def __init__(self, kind: Kind, payload: Union[str, BuildStatus]):
-            pods.
-        docker_host : str
-            The docker socket to use for building the image.
-            Must be a unix domain socket on a filesystem path accessible on the
-            node in which the build pod is running.
-        image_name : str
-            Full name of the image to build. Includes the tag.
-            Passed through to repo2docker.
-        git_credentials : str
-            Git credentials to use to clone private repositories. Passed
-            through to repo2docker via the GIT_CREDENTIAL_ENV environment
-            variable. Can be anything that will be accepted by git as
-            a valid output from a git-credential helper. See
-            https://git-scm.com/docs/gitcredentials for more information.
-        push_secret : str
-            Kubernetes secret containing credentials to push docker image to registry.
-        memory_limit
-            Memory limit for the docker build process. Can be an integer in
-            bytes, or a byte specification (like 6M).
-            Passed through to repo2docker.
-        memory_request
-            Memory request of the build pod. The actual building happens in the
-            docker daemon, but setting request in the build pod makes sure that
-            memory is reserved for the docker build in the node by the kubernetes
-            scheduler.
-        node_selector : dict
-            Node selector for the kubernetes build pod.
-        appendix : str
-            Appendix to be added at the end of the Dockerfile used by repo2docker.
-            Passed through to repo2docker.
-        log_tail_lines : int
-            Number of log lines to fetch from a currently running build.
-            If a build with the same name is already running when submitted,
-            only the last `log_tail_lines` number of lines will be fetched and
-            displayed to the end user. If not, all log lines will be streamed.
-        sticky_builds : bool
-            If true, builds for the same repo (but different refs) will try to
-            schedule on the same node, to reuse cache layers in the docker daemon
-            being used.
-        """
-        self.q = q
-        self.api = api
-        self.repo_url = repo_url
-        self.ref = ref
-        self.name = name
-        self.namespace = namespace
-        self.image_name = image_name
-        self.push_secret = push_secret
-        self.build_image = build_image
-        self.main_loop = IOLoop.current()
-        self.memory_limit = memory_limit
-        self.memory_request = memory_request
-        self.docker_host = docker_host
-        self.node_selector = node_selector
-        self.appendix = appendix
-        self.log_tail_lines = log_tail_lines
+    image_name = Unicode(help="Full name of the image to build. Includes the tag.")
 
-        self.stop_event = threading.Event()
-        self.git_credentials = git_credentials
+    git_credentials = Any(
```

Comment:
I've made it `CUnicode` since
https://github.com/jupyterhub/binderhub/blob/656204bb62d46be51fc7fb79cc5d7600b71a7614/binderhub/tests/test_build.py#L179-L182
passes a dict. Arguably it should be `Unicode` to avoid unexpected string conversions, but I think that can be done in a follow-up breaking PR if we want to.

Reply:
I think this was a bug in the test - an env var can't have a value that's a dict, and [this](https://github.com/jupyterhub/binderhub/blob/master/binderhub/build.py#L382) is how we use it:

```python
        if self.git_credentials:
            env.append(
                client.V1EnvVar(name="GIT_CREDENTIAL_ENV", value=self.git_credentials)
            )
```

So I think it does strictly need to be a string. This is type info revealing a bug!

Reply:
I've changed it to `Unicode`